internal/policygen/iam.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policygen
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/GoogleCloudPlatform/healthcare-data-protection-suite/internal/terraform"
+	"github.com/hashicorp/terraform/states"
+)
+
+type binding struct {
+	Type   string
+	ID     string
+	Role   string
+	Member string
+}
+
+func generateIAMBindingsPolicies(resources []*states.Resource, outputDir string) error {
+
+	bindings, err := bindings(resources)
+	if err != nil {
+		return err
+	}
+	log.Printf("Found %d bindings from input Terraform resources", len(bindings))
+	// TODO(xingao): Support google_*_iam_bindings.
+	// TODO(xingao): Generate policies.
+
+	return nil
+}
+
+func bindings(resources []*states.Resource) ([]*binding, error) {
+	typeToIDField := map[string]string{
+		"project":      "project",
+		"folder":       "folder",
+		"organization": "org_id",
+	}
+
+	var bindings []*binding
+	for t, idField := range typeToIDField {
+		instances, err := terraform.GetInstancesForType(resources, fmt.Sprintf("google_%s_iam_member", t))
+		if err != nil {
+			return nil, fmt.Errorf("get resource instances for type %q: %v", fmt.Sprintf("google_%s_iam_member", t), err)
+		}
+
+		for _, i := range instances {
+			if err := validate(i, idField, "role", "member"); err != nil {
+				return nil, err
+			}
+
+			bindings = append(bindings, &binding{
+				Type:   t,
+				ID:     i[idField].(string), // Type checked in validate()
+				Role:   i["role"].(string),
+				Member: i["member"].(string),
+			})
+		}
+	}
+	return bindings, nil
+}
+
+// validate checks the presense of mandatory fields and assert string type.
+func validate(instance map[string]interface{}, mandatoryFields ...string) error {
```

Comment:
I would suggest making `mandatoryFields` a `[]string`. This way, when calling the function, it's clearer that you're really working with two things (an instance, and a list of mandatory fields that it should contain).

Reply:
Done